pkg/chains/formats/slsa/internal/material/material_test.go

```diff
@@ -418,8 +420,114 @@ func TestAddImageIDToMaterials(t *testing.T) {
 		}
 		if tc.wantError == nil {
 			if diff := cmp.Diff(tc.want, mat, test.OptSortMaterial); diff != "" {
-				t.Errorf("materials(): -want +got: %s", diff)
+				t.Errorf("Materials(): -want +got: %s", diff)
 			}
 		}
 	}
 }
+
+func TestRemoveDuplicates(t *testing.T) {
+	tests := []struct {
+		name string
+		mats []slsa.ProvenanceMaterial
+		want []slsa.ProvenanceMaterial
+	}{{
+		name: "no duplicate materials",
+		mats: []slsa.ProvenanceMaterial{
+			{
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init",
+				Digest: slsa.DigestSet{
+					"sha256": "b963f6e7a69617db57b685893256f978436277094c21d43b153994acd8a01247",
+				},
+			}, {
+				URI: "gcr.io/cloud-marketplace-containers/google/bazel",
+				Digest: slsa.DigestSet{
+					"sha256": "010a1ecd1a8c3610f12039a25b823e3a17bd3e8ae455a53e340dcfdd37a49964",
+				},
+			}, {
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/sidecar-git-init",
+				Digest: slsa.DigestSet{
+					"sha256": "a1234f6e7a69617db57b685893256f978436277094c21d43b153994acd8a09567",
+				},
+			},
+		},
+		want: []slsa.ProvenanceMaterial{
+			{
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init",
+				Digest: slsa.DigestSet{
+					"sha256": "b963f6e7a69617db57b685893256f978436277094c21d43b153994acd8a01247",
+				},
+			}, {
+				URI: "gcr.io/cloud-marketplace-containers/google/bazel",
+				Digest: slsa.DigestSet{
+					"sha256": "010a1ecd1a8c3610f12039a25b823e3a17bd3e8ae455a53e340dcfdd37a49964",
+				},
+			}, {
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/sidecar-git-init",
+				Digest: slsa.DigestSet{
+					"sha256": "a1234f6e7a69617db57b685893256f978436277094c21d43b153994acd8a09567",
+				},
+			},
+		},
+	}, {
+		name: "same uri and digest",
+		mats: []slsa.ProvenanceMaterial{
+			{
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init",
+				Digest: slsa.DigestSet{
+					"sha256": "b963f6e7a69617db57b685893256f978436277094c21d43b153994acd8a01247",
+				},
+			}, {
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init",
+				Digest: slsa.DigestSet{
+					"sha256": "b963f6e7a69617db57b685893256f978436277094c21d43b153994acd8a01247",
+				},
+			},
+		},
+		want: []slsa.ProvenanceMaterial{
+			{
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init",
+				Digest: slsa.DigestSet{
+					"sha256": "b963f6e7a69617db57b685893256f978436277094c21d43b153994acd8a01247",
+				},
+			},
+		},
+	}, {
+		name: "same uri but different digest",
+		mats: []slsa.ProvenanceMaterial{
+			{
+				URI: "gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init",
```

Comment:
Thanks @jagathprakash for the comments !! 

1. We can have same uri but different digest (eg. if using two different commits from the same repo for different tasks).
2. Good question! If anything, we should combine the different digests instead of ignoring one completely. WDYT? I don't know how this might happen since we are fetching the digest so we would not selectively do this for one material and not for the other material. So I'm not sure if I should be worried about this. WDYT?
3. This, I agree. However, I think since the key is the same, the way we will merge it is by replacing the value in the map so it should not make much of a difference.

Reply:
Sounds good.